datacube/utils/documents.py

```diff
@@ -180,6 +180,33 @@ def add_uri_with_part(x):
             raise InvalidDocException(f"Failed to load {p}: {e}") from None
 
 
+def parse_doc_stream(
+    doc_stream: Sequence[tuple[str, str | bytes]],
+    on_error: Callable[[str, str | bytes], None] | None = None,
+    transform: Callable[[Mapping[str, Any]], Mapping[str, Any]] | None = None,
+) -> Generator[tuple[str, Mapping[str, Any] | None]]:
+    """
+    Replace doc bytes/strings with parsed dicts.
+
+    :param doc_stream: sequence of tuples consisting of uri and doc
+    :param on_error: Callback that gets the uri and doc as parameters
+    :param transform: if supplied, transforms the parsed document further
+
+    On output doc is replaced with python dict parsed from yaml, or with None
+    if parsing/transform error occurred.
```

Comment:
```suggestion
    Parse a stream of (filename, document body) tuples
    
    The document bodies are interpreted as either YAML or JSON depending on the filename suffix
    and turned into dictionary structures.
    
    If any error occurs while parsing, the ``on_error`` callback is run, and None is returned
    instead of a dictionary.

    :param doc_stream: sequence of tuples consisting of uri and document body
    :param on_error: error callback that gets the uri and doc as parameters
    :param transform: if given, transforms the parsed document

```

I don't know if this docstring is better or not. :)